lxd/db/query/dump.go

```diff
@@ -4,96 +4,116 @@ import (
+	builder.WriteString("PRAGMA foreign_keys=OFF;\n")
+	builder.WriteString("BEGIN TRANSACTION;\n")
 
-		tableDump, err := dumpTable(ctx, tx, table, schemas[table])
-		if err != nil {
-			return "", fmt.Errorf("Failed to dump table %s: %w", table, err)
-		}
+	// For each table, write the schema and optionally write the data.
+	for _, tableName := range tableNames {
+		builder.WriteString(tablesSchemas[tableName] + "\n")
 
-		dump += tableDump
+		if !schemaOnly {
+			tableData, err := getTableData(ctx, tx, tableName)
+			if err != nil {
+				return "", err
+			}
+
+			for _, stmt := range tableData {
+				builder.WriteString(stmt + "\n")
+			}
+		}
 	}
 
-	// Sequences (unless the schemaOnly flag is true)
+	// Sequences (unless the schemaOnly flag is true).
 	if !schemaOnly {
-		tableDump, err = dumpTable(ctx, tx, "sqlite_sequence", "DELETE FROM sqlite_sequence;")
+		builder.WriteString("DELETE FROM sqlite_sequence;\n")
+
+		tableData, err := getTableData(ctx, tx, "sqlite_sequence")
 		if err != nil {
 			return "", fmt.Errorf("Failed to dump table sqlite_sequence: %w", err)
 		}
 
-		dump += tableDump
+		for _, stmt := range tableData {
+			builder.WriteString(stmt + "\n")
+		}
 	}
 
-	// Commit
-	dump += "COMMIT;\n"
+	// Commit.
+	builder.WriteString("COMMIT;\n")
 
-	return dump, nil
+	return builder.String(), nil
 }
 
-// Return a map from table names to their schema definition, taking a full
-// schema SQL text generated with schema.Schema.Dump().
-func dumpParseSchema(schema string) map[string]string {
-	tables := map[string]string{}
-	for _, statement := range strings.Split(schema, ";") {
-		statement = strings.Trim(statement, " \n") + ";"
-		if !strings.HasPrefix(statement, "CREATE TABLE") {
-			continue
+// getTablesSchemas gets all the tables and their schema, as well as a list of table names in their default order from
+// the sqlite_master table.
+func getTablesSchemas(ctx context.Context, tx *sql.Tx) (map[string]string, []string, error) {
+	rows, err := tx.QueryContext(ctx, `SELECT name, sql FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' ORDER BY rowid`)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Could not get table names and their schema: %w", err)
+	}
+
+	defer rows.Close()
+
+	tablesSchemas := make(map[string]string)
+	var names []string
+	for rows.Next() {
+		var name string
+		var schema string
+		err := rows.Scan(&name, &schema)
+		if err != nil {
+			return nil, nil, fmt.Errorf("Could not scan table name and schema: %w", err)
+		}
+
+		names = append(names, name)
+
+		// Whether a table name is quoted or not can depend on if it was quoted when originally created, or if it
+		// collides with a keyword (and maybe more). Regardless, sqlite3 quotes table names in create statements when
+		// executing a dump. If the table name is already quoted, add the "IF NOT EXISTS" clause, else quote it and add
+		// the same clause.
+		isQuoted := strings.Contains(schema, fmt.Sprintf("TABLE %q", name))
```

Comment:
This section caught my eye. Why can we not do what line 89 does for all scenarios?

Why is setting `TABLE IF NOT EXISTS` dependent on quoting?

Reply:
My thinking here was that we have either a) `CREATE TABLE pets` or b) `CREATE TABLE "pets"`. 

If it is b, we don't need to alter the name but we need to change `TABLE` to `TABLE IF NOT EXISTS`. However if it is a) we need to wrap the name in quotes as well as add the `IF NOT EXISTS`. 

This was the only way I could think to do it with only one `strings.Replace` in each case.

Reply:
I see, makes sense now, thanks.